File: c++/valid_parentheses.cpp
#include <stack>
#include <string>
#include <iostream>

bool ValidParentheses(std::string string)
{
	if(string.length() % 2 != 0) return false;

	std::stack<char> stack;

	for(char character : string)
	{
		if(character == '(' || character == '[' || character == '{') stack.push(character);

		if(stack.empty())
		{
			if(character == ')' || character == ']' || character == '}') return false;
		}

		if(!stack.empty())
		{
			if(character == ')' && stack.top() == '(') stack.pop();
			else if(character == ')' && stack.top() != '(') return false;

			if(character == ']' && stack.top() == '[') stack.pop();
			else if(character == ']' && stack.top() != '[') return false;

			if(character == '}' && stack.top() == '{') stack.pop();
			else if(character == '}' && stack.top() != '{') return false;
		}
	}

	return stack.empty();
}

std::string string = {"([}}])"};

int main()
{
	std::cout << std::boolalpha << ValidParentheses(string) << "\n";
}

File: go/valid_parentheses.go
package main

import "fmt"

func validParentheses(str string) bool {
	if len(str)%2 != 0 {
		return false
	}

	stack := []rune{}

	for _, character := range str {
		if character == '(' || character == '[' || character == '{' {
			stack = append(stack, character)
		}

		if len(stack) == 0 {
			if character == ')' || character == ']' || character == '}' {
				return false
			}
		}

		if len(stack) > 0 {
			if character == ')' && stack[len(stack)-1] == '(' {
				stack = stack[:len(stack)-1]
			} else if character == ')' && stack[len(stack)-1] != '(' {
				return false
			}

			if character == ']' && stack[len(stack)-1] == '[' {
				stack = stack[:len(stack)-1]
			} else if character == ']' && stack[len(stack)-1] != '[' {
				return false
			}

			if character == '}' && stack[len(stack)-1] == '{' {
				stack = stack[:len(stack)-1]
			} else if character == '}' && stack[len(stack)-1] != '{' {
				return false
			}
		}
	}

	return len(stack) == 0
}

func main() {
	str1 := "[{[]}]"
	str2 := "][}}]["
	str3 := "])}{(["
	str4 := "([{}])"

	fmt.Println(validParentheses(str1))
	fmt.Println(validParentheses(str2))
	fmt.Println(validParentheses(str3))
	fmt.Println(validParentheses(str4))
}
